scripts/extract_historical_product_data.pl

```diff
@@ -0,0 +1,335 @@
+			$query_ref->{$field} += 0;
+		}
+		# Multiple values separated by commas
+		elsif ($query_ref->{$field} =~ /,/) {
+			my @tagids = split(/,/, $query_ref->{$field});
+
+			if ($not) {
+				$query_ref->{$field} = {'$nin' => \@tagids};
+			}
+			else {
+				$query_ref->{$field} = {'$in' => \@tagids};
+			}
+		}
+		elsif ($not) {
+			$query_ref->{$field} = {'$ne' => $query_ref->{$field}};
+		}
+	}
+
+	use Data::Dumper;
+	print STDERR "MongoDB query:\n" . Dumper($query_ref);
+
+	my $socket_timeout_ms = 2 * 60000;    # 2 mins, instead of 30s default, to not die as easily if mongodb is busy.
+
+	# Collection that will be used to iterate products
+	# TODO: we should query both the products and obsolete products collections
+	my $products_collection = get_products_collection({obsolete => 0, timeout => $socket_timeout_ms});
+
+	my $products_count = "";
+
+	eval {
+		$products_count = $products_collection->count_documents($query_ref);
+
+		print STDERR "$products_count documents retrieved from MongoDB\n";
+	};
+
+	my $cursor = $products_collection->query($query_ref)->fields({_id => 1, code => 1, owner => 1});
+	$cursor->immortal(1);
+
+	while (my $product_ref = $cursor->next) {
+		push @codes, $product_ref->{code};
+	}
+}
+
+# load data needed to analyze and enrich products
+if ($analyze_and_enrich_product_data) {
+	load_data();
+}
+
+# Print the header
+my $years = "";
+for (my $year = $min_year; $year <= $max_year; $year++) {
+	$years .= "\t" . $year;
+}
+
+print "code" . "\t" . "found" . "\t" . "url" . $years . "\n";
+
+# Go through all products
+my $products = 0;
+my $found_products = 0;
+
+foreach my $code (@codes) {
+
+	$products++;
+
+	my $productid = $code;
+	my $path = product_path_from_id($productid);
+	my $found = 0;
+
+	# We will go through revision one by one, and for the requested field, we will
+	# store the value we have at the beginning of each year
+	my %value_per_year = ();
+	my $url = "";
+
+	# Go through all product revisions
+	my $changes_ref = retrieve("$BASE_DIRS{PRODUCTS}/$path/changes.sto");
+
+	# If we don't have a changes.sto file, the product is not in the database
+	if (defined $changes_ref) {
+		$found = 1;
+		$found_products++;
+
+		# Go through all revisions, keep the latest value of all fields
+
+		my %deleted_values = ();
+		my $previous_product_ref = {};
+		my $revs = 0;
+
+		foreach my $change_ref (@{$changes_ref}) {
+			$revs++;
+			my $rev = $change_ref->{rev};
+			if (not defined $rev) {
+				$rev = $revs;    # was not set before June 2012
+			}
+
+			my $product_ref = retrieve("$BASE_DIRS{PRODUCTS}/$path/$rev.sto");
+
+			if (defined $product_ref) {
+
+				my $value;
+
+				# Make sure we have a lc field
+				if ((not defined $product_ref->{lc}) and (defined $product_ref->{lang})) {
+					$product_ref->{lc} = $product_ref->{lang};
+				}
+
+				# Determine the year of the revision, using the UNIX timestamp last_modified_t
+				my $year = (localtime($product_ref->{last_modified_t}))[5] + 1900;
```

Comment:
Instead of computing values for each revision, while at the end you only keep value for last revision of the year, why not jump until last revision of this year ? 

That is first, get a map of (last) product_ref by year, and then process values ? (interesting to skip calls to analyze_and_enrich_product_data)

Reply:
Good point, changed.